pallets/stellar-relay/src/validation.rs

```diff
@@ -0,0 +1,224 @@
+use frame_support::{ensure, BoundedVec};
+use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
+
+use substrate_stellar_sdk::{
+	compound_types::UnlimitedVarArray,
+	network::Network,
+	types::{NodeId, ScpEnvelope, ScpStatementPledges, Value},
+	Hash,
+};
+
+use crate::{
+	pallet::{verify_signature, Config},
+	types::{OrganizationOf, ValidatorOf},
+	Error, NewValidatorsEnactmentBlockHeight, OldOrganizations, OldValidators, Organizations,
+	Pallet, Validators,
+};
+
+/// Returns a map of organizationID to the number of validators that belongs to it
+fn validator_count_per_org<T: Config>(
+	validators: &BoundedVec<ValidatorOf<T>, T::ValidatorLimit>,
+) -> BTreeMap<T::OrganizationId, u32> {
+	let mut validator_count_per_organization_map = BTreeMap::<T::OrganizationId, u32>::new();
+
+	for validator in validators.iter() {
+		validator_count_per_organization_map
+			.entry(validator.organization_id)
+			.and_modify(|e| {
+				*e += 1;
+			})
+			.or_insert(1);
+	}
+
+	validator_count_per_organization_map
+}
+
+/// Builds a map used to identify the targeted organizations
+fn targeted_organization_map<T: Config>(
+	envelopes: &UnlimitedVarArray<ScpEnvelope>,
+	validators: &BoundedVec<ValidatorOf<T>, T::ValidatorLimit>,
+	orgs_length: usize,
+) -> Result<BTreeMap<T::OrganizationId, u32>, Error<T>> {
+	// Find the validators that are targeted by the SCP messages
+	let targeted_validators = validators
+		.iter()
+		.filter(|validator| {
+			envelopes.get_vec().iter().any(|envelope| {
+				envelope.statement.node_id.to_encoding() == validator.public_key.to_vec()
+			})
+		})
+		.collect::<Vec<&ValidatorOf<T>>>();
+
+	// A map is used to avoid duplicates and simultaneously track the number of validators
+	// that were targeted
+	let mut targeted_organization_map = BTreeMap::<T::OrganizationId, u32>::new();
+	for validator in targeted_validators {
+		targeted_organization_map
+			.entry(validator.organization_id)
+			.and_modify(|e| {
+				*e += 1;
+			})
+			.or_insert(1);
+	}
+
+	// Count the number of distinct organizations that are targeted by the SCP messages
+	let targeted_organization_count = targeted_organization_map.len();
+
+	// Check that the distinct organizations occurring in the validator structs related to
+	// the externalized messages are more than 2/3 of the total amount of organizations in
+	// the tier 1 validator set.
+	// Use multiplication to avoid floating point numbers.
+	ensure!(
+		targeted_organization_count * 3 > orgs_length * 2,
+		Error::<T>::InvalidQuorumSetNotEnoughOrganizations
+	);
```

Comment:
I would move this part also to the `check_targets()` function. It's not wrong here but it's better fitting to the other one, especially if we rename it to `check_for_valid_quorum_set` as this part of related to that.

Reply:
updated